packages/redux-dynostore-core/src/deepMerge.js

```diff
@@ -0,0 +1,67 @@
+/**
```

Comment:
I did try most of the main players initially ([`deepmerge`](https://www.npmjs.com/package/deepmerge) and [`merge-deep`](https://www.npmjs.com/package/merge-deep)), and tried [copying someone else's example](https://davidwalsh.name/javascript-deep-merge), but none of these quite fit some of the requirements I had, particularly around dealing with Redux best practices.

The first issue was that they did not return referentially equal values when the state, or sub-sections of the state, did not actually change, they just blindly clone everything all the way down the tree.  `deepmerge` did have an option to not clone (which is `deprecated` in their docs), but this would change the behaviour to mutate the `target` instead, which is also a Redux no-no.

The second issue was how they dealt with arrays.  Consider these two objects:

```js
const state1 = {
  items: [{ key: 'value1' }, { key: 'value2' }]
}

const state2 = {
  items: [{ key: 'value1' }, { key: 'value3' }]
}
```

Generally, the merge algorithm's default behaviour is to concat these arrays, so the output becomes:

```js
const result = {
  items: [{ key: 'value1' }, { key: 'value2' }, { key: 'value1' }, { key: 'value3' }]
}
```

This does not work in our scenario as the reason the array was found in the previous state is most likely because the it was the second reducers state to begin with (carried over from the previous update).  The consequence is that the array keeps growing after every dispatch.

Instead, my deep merge merges items found at the same index in both array, and adds any additional items found in either array.  The referential equality mentioned above means the if the items were the same (or contained the same values internally), the item does not change at all, otherwise a new item is used with the updated values.  This means the second reducer can freely update it's values in place, just like it would expect in regular Redux usage.

The flip side is if you genuinely had 2 reducers adding new items to the same array in a single dispatch, the two new items would be merged together.  If they are adding different keys item in the array this will likely be what you want anyway, but if they use the same keys this might cause some confusing behaviour.  I would argue that if your app is set up this way, there is already some confusing behaviour occurring and you should not have 2 reducers in the first place.

Reply:
Thanks for the explanation.